WIP: Reactions frontend

arpith opened the pull request:
This PR replaces the placeholder "clipboard" button with a reaction button.
This is done on any message that can't be edited.
When clicked, a popover with a search bar and a list of emojis is displayed.
If the right sidebar is collapsed (the viewport is small), the popover is placed
to the left of the button.
Focus is set to the search bar. Typing in the search bar filters emojis.

Emojis with which the user has reacted to this message are highlighted.
Clicking them sends an API request to remove that reaction.
Clicking on non-highlighted emojis sends an API request to add a reaction.
When the popover loses focus it is closed.

The frontend listens for reaction events. When an add-reaction event is
received, the emoji is displayed at the bottom of the message with a
count initialized to 1. If there was an existing reaction to the message with
the same emoji, the count is incremented.

Old messages fetched from the server contain reactions.
They are displayed (along with title and count) at the bottom
of each message.

When clicking the emoji reaction at the bottom of the message, if the
user has already reacted with that emoji to this message, the reaction
is removed and the count is decremented. Otherwise, a reaction is added
and the count is incremented.

Hovering over the emoji reaction at the bottom of the message displays
a list of users who have reacted with this emoji along with the
emoji name.

Hovering over the emoji reactions at the bottom of the message displays
a button to add a reaction.

smarx commented:
*Automated message from Dropbox CLA bot*

@arpith, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
@arpith thanks for working on this!  Obviously this is a WIP, but here are a few comments:
* We should replace the clipboard to view source with the "add emoji" icon -- that's always been the plan for what goes in that spot.  (View source is still available in the caret menu).
* There's a bunch of lint/test failures to address :).
* Actually picking an emoji doesn't work :)
* The scrollbar for the area looks ugly on Linux if you type in a few characters (e.g. `alie`.  
* It's great that the typeahead works!

arpith commented:
Thanks for your feedback @timabbott! I've replaced the clipboard with the add emoji icon, fixed the lint errors, and now actually picking an emoji displays the reaction with a count 😄 I've made some fixes for the scroll issue, wondering if I can use the screenshot testing tool for this.

timabbott commented on static/js/ui.js:
```diff
@@ -117,12 +117,7 @@ function message_hover(message_row) {
-    } else {
-        message_row.find(".edit_content").html('<i class="icon-vector-file-text-alt edit_content_button" data-msgid="' + id + '"></i>');
-    }
+    message_row.find(".reaction_button").html('<i class="icon-vector-smile edit_content_button" data-msgid="' + id + '"></i>');
```
So, we probably still want it to show the pencil for your own, editable messages; it's just for other users' messages (i.e. the `file` icon case) where we want to have the reaction icon.

timabbott commented:
I'm not sure what you mean by screenshot testing tool -- do you mean Casper?  We don't really use the screenshots generated by it for anything...

timabbott commented:
I played around with this and it still feels pretty rough:
* It probably makes sense to put a magnifying class next to the search box (or something) to indicate what the text box is for visually.
* The keyboard focus should go into the "input emoji" box when you click the "add emoji" icon.
* The autocomplete doesn't seem to work properly (e.g. I type in "cow", and see nothing -- because it doesn't spacially hide the emoji that don't match):
![image](https://cloud.githubusercontent.com/assets/2746074/21170332/1a6afee0-c179-11e6-9fdf-dd2bb9c13eb7.png)
* In a narrow window, where the right sidebar is hidden, the right half of the popover gets hidden.  It should shift to popping right in that case I guess?
* It's hard to close the add-emoji popover once you open it; for me it doesn't close when you react (which it definitely should) or via escape/etc.
* there's random gaps in the emoji list (see screenshot below)
* When you mouse over the current already-sent emoji, it should show you who sent the various emoji, not just which one it was.
![image](https://cloud.githubusercontent.com/assets/2746074/21170299/c9c8c2a6-c178-11e6-8c2d-fab4813bf5a9.png)

arpith commented:
Thanks for your comments @timabbott :) I've made the fixes except for displaying realm emojis

1. Added a magnifying glass to the popover
2. Clicking on the reaction button takes focus to the filter/search field
3. Filtered (non-matching) emoji are spatially hidden
4. The popover opens to the left in a narrow window
5. Clicking on an emoji / outside the popover closes the popover. A single escape takes focus from the text field, a second one closes the popover
6. Random gaps due to white emojis are fixed (blue emojis merged)
7. Titles of message reaction now show who has reacted

Next step - using reaction data fetched via get_old_messages!

timabbott commented:
Cool!  This looks great from a functionality perspective!  A few notes on little UI polish issues:
* The emoji symbol in the message body doesn't highlight on hover like the caret does.
* Clicking on an emoji reaction to delete it pops open the compose box; I think you're missing a preventDefault or stopPropagation there.
* When hovering over an emoji reaction, the message should include both who reacted and what the reaction was.  Slack's formatting for this seems pretty reasonable to copy.
* The emoji reactions should probably be intended to align with the start of message bodies, not with the sender name.
* It appears with this CSS, every message has a space under it for emoji reactions.  That vertical space should only be present if there are actual emoji reactions in the current message.
* There should probably be a visible-on-hover-over-the-emoji-reactions-row offer to add your own (Slack does this well).
* The emoji picker scrollbar should probably stay on the right (or maybe just be invisible) in the case that there aren't enough emoji to pick in this screenshot:
![image](https://cloud.githubusercontent.com/assets/2746074/21398711/28d04942-c75d-11e6-80d5-ca99081b9a06.png)

If you have limited time, I'd focus on the Javascript/logic items; @brockwhittaker will probably be faster on the CSS optimizations.  

Now going to take a quick look at the code as well.

timabbott commented on static/js/click_handlers.js:
```diff
@@ -83,6 +83,8 @@ $(function () {
         toggle_star(rows.id($(this).closest(".message_row")));
     });
 
+    $("#main_div").on("click", ".message_reaction", reactions.message_reaction_on_click);
```
Stylistically, this is probably where the preventDefault() should go.

arpith replied:
Done :)

timabbott commented on static/js/popovers.js:
```diff
@@ -43,6 +44,51 @@ function show_message_info_popover(element, id) {
+            title:     "",
+            content:   templates.render('reaction_popover_content', args),
+            trigger:   "manual"
+        });
```
empirically, this seems to always pop left; is that your intent?

arpith replied:
Fixed!

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,194 @@
+    get_reactions_for_message: get_reactions_for_message,
+    update_reaction_store: update_reaction_store,
+    message_reaction_on_click: message_reaction_on_click
+};
```
Can you redo the exports in the same style as our other JavaScript modules, where what is exported is declared in the function declaration?

arpith replied:
Fixed :)

timabbott commented on static/js/message_store.js:
```diff
@@ -409,6 +409,7 @@ function process_result(messages, opts) {
     activity.process_loaded_messages(messages);
     stream_list.update_streams_sidebar();
     pm_list.update_private_messages();
+    reactions.update_reaction_store();
```
Doesn't this need to pass the list of messages?

arpith replied:
Removed :) (there is no store to update now)

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,194 @@
+            operation = 'add';
+        }
+    }
+    return operation;
```
This could probably be clearer if it eliminated the `operation` variable and just used `return` statements.

arpith replied:
Done :)

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,194 @@
+    var message_id = event.message_id;
+    var stored_reactions = reactions_store.get(message_id);
+    if (!stored_reactions.has(emoji_name)) {
+      return;
```
the indentation here is off

arpith replied:
Fixed :)

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,194 @@
+var reactions = (function () {
+var exports = {};
+
+var reactions_store = new Dict();
```
Why do you need this object?  It seems like you could just keep the reaction data in the relevant message_store object for the message; is there an advantage to having this extra dict off to the side?  It seems like it adds some extra accounting complexity...

arpith replied:
Removed :)

arpith commented:
Thanks for your comments @timabbott :) I've made the changes:

1. Hovering over the reaction button changes its color to blue
2. Clicking on the reaction at the bottom of the message doesn't open the compose box
3. Hovering over an emoji reaction displays who reacted as well as the emoji name
4. Emoji reactions are now aligned with the start of the message body
5. Removed the vertical space below the message when there are no reactions
6. Hovering over the message reactions displays an "add reaction button", which, when clicked, opens the popover
7. Forced the scrollbar to stay right

timabbott commented:
A few bugs:
* You should URL-encode the emoji names, similar to how we do in various places in subs.js with stream names, so that emoji like `+1` work.
* The UI for clicking on the emoji reactions is broken (throws a traceback consistently for me), I think because of some problem with storing the message ID in the title.  We should be using `data-message-id` rather than `title` to store this anyway.

timabbott commented:
This should fix the second problem:
```
     $("#main_div").on("click", ".message_reaction", function (e) {
         e.stopPropagation();
+        var row = current_msg_list.get_row(rows.id($(this).closest(".message_row")));
         var emoji_name = $(this).attr('data-emoji-name');
-        var message_id = $(this).parent()[0].title;
+        var message_id = rows.id(row);
         reactions.message_reaction_on_click(message_id, emoji_name);
     });
```

timabbott commented:
These changes fix the `+1` issue, mostly (it still doesn't display properly?):
```
diff --git a/static/js/reactions.js b/static/js/reactions.js
index c9213ff..36b593a 100644
--- a/static/js/reactions.js
+++ b/static/js/reactions.js
@@ -3,7 +3,7 @@ var exports = {};
 
 function send_reaction_ajax(message_id, emoji_name, operation) {
     var args = {
-        url: '/json/messages/' + message_id + '/emoji_reactions/' + emoji_name,
+        url: '/json/messages/' + message_id + '/emoji_reactions/' + encodeURIComponent(emoji_name),
         data: {},
         success: function () {},
         error: function (xhr) {
diff --git a/zproject/urls.py b/zproject/urls.py
index 4cabd0a..8537063 100644
--- a/zproject/urls.py
+++ b/zproject/urls.py
@@ -206,7 +206,7 @@ v1_api_and_json_patterns = [
     # reactions -> zerver.view.reactions
     # PUT adds a reaction to a message
     # DELETE removes a reaction from a message
-    url(r'^messages/(?P<message_id>[0-9]+)/emoji_reactions/(?P<emoji_name>[0-9a-zA-Z.\-_]+(?<![.\-_]))$',
+    url(r'^messages/(?P<message_id>[0-9]+)/emoji_reactions/(?P<emoji_name>[+0-9a-zA-Z.\-_]+(?<![.\-_]))$',
         rest_dispatch,
         {'PUT': 'zerver.views.reactions.add_reaction_backend',
          'DELETE': 'zerver.views.reactions.remove_reaction_backend'}),
```

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,176 @@
+        error: function (xhr) {
+          var response = channel.xhr_error_message("Error sending reaction", xhr);
+          blueslip.error(response);
+        }
```
This isn't indented properly and throws an exception if you try a nonexisting emoji.

arpith replied:
Fixed!

timabbott commented on static/templates/single_message.handlebars:
```diff
@@ -60,6 +61,7 @@
         </div>
         <div class="message_expander message_length_controller" title="{{t 'See the rest of this message' }}">{{t "[More...]" }}</div>
         <div class="message_condenser message_length_controller" title="{{t 'Make this message take up less space on the screen' }}">{{t "[Condense this message]" }}</div>
+        <div class="message_reactions {{msg/id}}">{{ partial "message_reactions" }}</div>
```
Remove the ` {{msg/id}}` class -- you probably don't want the message ID as a class :)

arpith replied:
Done :)

timabbott commented on static/templates/message_reaction.handlebars:
```diff
@@ -0,0 +1,4 @@
+<div class="message_reaction" title="{{this.title}}" data-emoji-name="{{this.emoji_name}}">
+  <div class="emoji emoji-{{this.emoji_name}}" />
+  <div class="message_reaction_count">{{this.count}}</div>
+</div>
```
Can this just be inlined into message_reactions.handlebars?  Seems like we don't need more little files to manage this...

arpith replied:
message_reaction.handlebars is used in add_reaction() (in reactions.js) to append new reactions to the message reactions div.

message_reactions.handlebars is used to generate the message reactions div on first load. This could be inlined into single_message.handlebars, but that file is big enough :)

timabbott replied:
makes sense, we can keep it as-is then.

arpith replied:
cool :)

timabbott commented on static/styles/reactions.css:
```diff
@@ -0,0 +1,106 @@
+    border: thin solid white;
+    border-radius: 0.5em;
+}
+.reaction-popover .reacted {
```
There should be a newline between CSS blocks to match our standard style.

arpith replied:
Fixed!

arpith commented on static/templates/single_message.handlebars:
```diff
@@ -38,7 +38,8 @@
             </div>
-            <span class="edit_content"></span>
+            <span class="edit_content">
+            </span>
```
Removed :)

arpith commented on static/templates/reaction_popover_content.handlebars:
```diff
@@ -0,0 +1,14 @@
+    <i class="icon-vector-search"></i>
+    <input class="reaction-popover-filter" type="text" autofocus placeholder={{t 'Search' }} />
+  </div>
+  <div class="reaction-popover-emoji-map" title={{message_id}}>
```
Removed :)

timabbott commented:
You've got some new linter failures :)

arpith commented on static/styles/reactions.css:
```diff
@@ -0,0 +1,106 @@
+}
+.popover-content {
+    padding: 0;
+}
```
Fixed :)

arpith commented on static/js/reactions.js:
```diff
@@ -0,0 +1,176 @@
+    var emoji_name = event.emoji_name;
+    var message_id = event.message_id;
+    event.user.id = event.user.user_id;
+    message_store.get(message_id).reactions.push(event);
```
Re: the first comment, messages in message_store have a reactions property that is a list of tuples (emoji_name, user_id). This is what get_old_messages sends from the server, a list of individual reactions (reactions in the database are unique on message_id, emoji_name and user_id). 
Adding a new reaction (emoji_name, user_id) to this list is what we want to do.

arpith replied:
Fixed the inconsistent state / titles :)

arpith commented on static/js/reactions.js:
```diff
@@ -0,0 +1,176 @@
+        user_ids.unshift(page_params.user_id);
+    }
+    return user_ids.map(full_name).join(', ') + ' reacted with :' + emoji_name + ':';
+}
```
Fixed :)

arpith commented:
Thanks for your comments @timabbott :) I've made the changes:

1. emoji names are url-encoded 
2. Switch to data-message-id from using message_id as class/title
3. Use data-message-id in message reaction click handler
4. Used changes to fix +1 issue, as well as fix display of +1
5. Fix indentation, don’t send reaction ajax request if emoji does not exist
6. Fix display of +1
7. Fix bug where sending a new message caused reactions on previous message to disappear
8. Added “add emoji reaction” option in actions popover for messages sent by user
9. Made reactions smaller
10. Set padding above and below emoji to be same, except for emoji inconsistencies.
11. Made reaction counts darker
12. Zulip emoji works now, set up ground work for realm emoji as well
13. Moved .popover-content styling into .reaction-popover
14. Add newline between css blocks in stylesheet
15. Remove change made to edit_content span
16. Switch from message_id as class to data-message-id
17. Fixed bug with title (splice)
18. Clarified the reactions data structure used by messages in message_store
19. Make state consistent when adding / removing reactions with two users 
20. Switch from message_id as title to data-message-id
21. Can’t move code from message_reaction.handlebars to message_reactions.handlebars because message_reactions.handlebars is used in single_message.handlebars, and message_reaction.handlebars is used to append reactions in add_reaction()
22. Fixed lint failures

I've also fixed the following bugs:
1. (array.splice() takes two args, not one): Add two reactions, remove them by clicking on them in the order they were added. The second one’s user list is empty and so an attempt is made to add it again, instead of removing it. Doing this in the opposite order works fine.
2. (add .reactions in message_store if undefined) When sending a message, the server event received has a reactions property (empty array), but by the time it reaches message_store this property is missing.
3. Clicking on reaction button below message in small windows opens popover to left (should be right)
4. (splice takes two args) Titles for reactions where user is the first in the list didn’t include the other users.
5. +1 doesn’t display in popover when it has been used by user (displays under message correctly)
6. Adding a reaction by user A and then adding the same reaction by user B results in reaction with count = 2 in user B’s page, but duplicated reaction (both with count 1) in user A’s page
7. Titles when more than one user has reacted should be “userA and userB”, not “userA, userB”
8. Removing reaction doesn’t change title
9. Adding a reaction in private / pm-with narrow doesn’t display the reaction (moving to home narrow displays it). 
10. $().data() converts emoji names that are numbers to numbers (not strings)

timabbott commented:
I found an additional bug, actually in the backend; I'm going to merge this to fix it:

```
commit 97d04a31cec726655d130f28829f4fc05f248e3f
Author: Tim Abbott <tabbott@zulipchat.com>
Date:   Fri Dec 30 17:45:16 2016 -0800

    reactions: Fix check for valid emoji.
    
    The previous logic did not correctly handle the case where there were
    realm emoji (non-realm emoji would be inaccessible).

diff --git a/zerver/views/reactions.py b/zerver/views/reactions.py
index a69a595..e2d1792 100644
--- a/zerver/views/reactions.py
+++ b/zerver/views/reactions.py
@@ -13,6 +13,13 @@ from zerver.lib.request import JsonableError
 from zerver.lib.response import json_success
 from zerver.models import Reaction, UserProfile
 
+def check_valid_emoji(realm, emoji_name):
+    if emoji_name in set(realm.get_emoji().keys()):
+        return
+    if emoji_name in emoji_list:
+        return
+    raise JsonableError(_("Emoji '%s' does not exist" % (emoji_name,)))
+
 @has_request_variables
 def add_reaction_backend(request, user_profile, message_id, emoji_name):
     # type: (HttpRequest, UserProfile, int, Text) -> HttpResponse
@@ -21,9 +28,7 @@ def add_reaction_backend(request, user_profile, message_id, emoji_name):
     # cannot see the message (e.g. for messages to private streams).
     message = access_message(user_profile, message_id)[0]
 
-    existing_emojis = set(message.sender.realm.get_emoji().keys()) or set(emoji_list)
-    if emoji_name not in existing_emojis:
-        raise JsonableError(_("Emoji '%s' does not exist" % (emoji_name,)))
+    check_valid_emoji(message.sender.realm, emoji_name)
 
     # We could probably just make this check be a try/except for the
     # IntegrityError from it already existing, but this is a bit cleaner.
@@ -44,9 +49,7 @@ def remove_reaction_backend(request, user_profile, message_id, emoji_name):
     # cannot see the message (e.g. for messages to private streams).
     message = access_message(user_profile, message_id)[0]
 
-    existing_emojis = set(message.sender.realm.get_emoji().keys()) or set(emoji_list)
-    if emoji_name not in existing_emojis:
-        raise JsonableError(_("Emoji '%s' does not exist" % (emoji_name,)))
+    check_valid_emoji(message.sender.realm, emoji_name)
 
     # We could probably just make this check be a try/except for the
     # IntegrityError from it already existing, but this is a bit cleaner.
```

timabbott commented:
Other than that, and some unfortunate fallout of my moving where emoji are served from (which I'll deal with), the only bug I can find is that someone else adding an emoji reaction above your place in the message view causes a scroll jolt, which we'll need to figure out how to fix.

timabbott commented on static/js/popovers.js:
```diff
@@ -170,13 +244,25 @@ function message_info_popped() {
 
+exports.hide_reactions_popover = function () {
+    if (reaction_popped()) {
+        $('.popover').remove();
```
This selector is not scoped properly

timabbott replied:
I think we can just delete this, but only after fixing the fact that the  `current_message_reactions_popover_elem` gets deleted once you unhover the message.

timabbott commented on static/js/popovers.js:
```diff
@@ -280,6 +366,19 @@ exports.register_click_handlers = function () {
+
+
+    $("body").on("click", ".reaction_button", function (e) {
+        var msgid = $(e.currentTarget).data('msgid');
```
We should rename this to data-message-id to match the style for our other data objects.

timabbott replied:
I guess we need to bulk-rename all the actions popover ones, so we can do this later.

timabbott commented on static/js/reactions.js:
```diff
@@ -0,0 +1,198 @@
+    }
+    send_reaction_ajax(message_id, emoji_name, operation);
+    popovers.hide_all();
+}
```
This could share code with the previous function

timabbott commented on static/js/message_list_view.js:
```diff
@@ -783,7 +793,11 @@ MessageListView.prototype = {
+        message_container.msg.message_reactions = msg_reactions;
+        var msg_to_render = _.extend(message_container, {
+            table_name: this.table_name,
+        });
```
These 3 identical blocks should be refactored into a function.

timabbott commented:
I fixed the popovers issues noted above as ec3e5e844a83c04319f039ba339defe46059d00f and merged this.  Huge huge thanks for your work on this @arpith!!

I'm going to open a follow-up issue for the refactors suggested above.

arpith commented:
Thanks @timabbott :)
